cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js

```diff
@@ -272,6 +277,41 @@ class HydratorPlusPlusTopPanelCtrl {
     };
   }
 
+  fetchMacros() {
+    this.macrosMap = [];
+    let nodes = this.HydratorPlusPlusConfigStore.getNodes();
+
+    for (let i = 0; i < nodes.length; i++) {
+      let properties = this.myHelpers.objectQuery(nodes[i], 'plugin', 'properties');
+      let backendProperties = this.myHelpers.objectQuery(nodes[i], '_backendProperties');
+      for (let prop in properties) {
+        if (properties.hasOwnProperty(prop) && backendProperties[prop].macroSupported) {
+          let propValue = properties[prop];
+          if (propValue.substring(0, 2) === '${' && propValue[propValue.length - 1] === '}') {
```

Comment:
I don't think this algorithm is correct. A macro can have nested macros, so this algorithm will fail.. We should check the backend code, how they interpret macro, and replicate here. Should add a FIXME though.. ideally we shouldn't duplicate this logic in the FE